core/src/main/scala/org/apache/spark/sql/execution/columnar/encoding/ColumnDeltaEncoder.scala

```diff
@@ -561,34 +546,27 @@ final class ColumnDeltaEncoder(val hierarchyDepth: Int) extends ColumnEncoder {
         (numNullWords << 3) /* header */ +
         // round positions to nearest word as done by writeHeader
         4 /* numBaseRows */ + 4 /* numPositions */ + (numDeltas << 2) + 7) >> 3) << 3) +
-        realEncoder.encodedSize(encoderCursor, realEncoder.columnBeginPosition + dataOffset)))
+        realEncoder.encodedSize(encoderCursor, dataBeginPosition)))
     realEncoder.setSource(buffer, releaseOld = false)
     val columnBytes = allocator.baseObject(buffer)
     var cursor = allocator.baseOffset(buffer)
     // write the header including positions
     cursor = writeHeader(columnBytes, cursor, numNullWords, numBaseRows,
-      positionsArray, startIndex, endIndex)
+      positionsArray, numDeltas)
 
     // write any internal structures (e.g. dictionary)
     cursor = realEncoder.writeInternals(columnBytes, cursor)
```

Comment:
Can't we call realEncoder.finish() here, which will write any internal structures ?

Reply:
Avoids one additional copy in finish. Even with finish another methods will be needed to determine the boundary where data starts to create the new delta layout as mentioned above (see the writeHeader method that also writes the updated positions).